pilot/pkg/xds/endpoint_builder.go

```diff
@@ -419,16 +425,21 @@ func (c *mtlsChecker) computeForEndpoint(ep *model.IstioEndpoint) {
 	if drMode := c.mtlsModeForDestinationRule(ep); drMode != nil {
 		switch *drMode {
 		case networkingapi.ClientTLSSettings_DISABLE:
-			c.mtlsDisabledHosts[lbEpKey(ep.EnvoyEndpoint)] = struct{}{}
-			return
+			if !c.sniDNAT {
+				c.mtlsDisabledHosts[lbEpKey(ep.EnvoyEndpoint)] = struct{}{}
+				return
+			}
+			// ClientTLSSettings shouldn't affect SNI DNAT clusters but we still check PeerAuthn/tlsMode label.
 		case networkingapi.ClientTLSSettings_ISTIO_MUTUAL:
-			// don't mark this EP disabled, even if PA or tlsMode meta mark disabled
+			// The client is forcing ISTIO_MUTUAL; don't bother checking PeerAuthentication or tlsMode labels.
+			// This setting won't actually apply on the SNI-DNAT clusters, but we still allow the setting as
+			// an escape hatch to avoid the filtering here.
 			return
 		}
 	}
 
-	// if endpoint has no sidecar or explicitly tls disabled by "security.istio.io/tlsMode" label.
-	if ep.TLSMode != model.IstioMutualTLSModeLabel {
+	// the endpoint must either be a part of istio or have the securty.istio.io/tlsMode: istio label
+	if !ep.IsIstioGateway() && ep.TLSMode != model.IstioMutualTLSModeLabel {
```

Comment:
Why checking IsGateway here? IIRC， tls mode is always set on gateway

Reply:
I don't think that's true from my testing. I don' think checking "is gateway" is a good idea, more of an experiment so far.